src/contracts/middleware/BLSOperatorStateRetriever.sol

```diff
@@ -96,36 +96,44 @@ contract BLSOperatorStateRetriever {
+    ) external view returns (CheckSignaturesIndices memory) {
         IStakeRegistry stakeRegistry = registryCoordinator.stakeRegistry();
         CheckSignaturesIndices memory checkSignaturesIndices;
 
         checkSignaturesIndices.nonSignerQuorumBitmapIndices = registryCoordinator.getQuorumBitmapIndicesByOperatorIdsAtBlockNumber(referenceBlockNumber, nonSignerOperatorIds);
         checkSignaturesIndices.totalStakeIndices = stakeRegistry.getTotalStakeIndicesByQuorumNumbersAtBlockNumber(referenceBlockNumber, quorumNumbers);
         
-        checkSignaturesIndices.nonSignerStakeIndices = new uint32[][](nonSignerOperatorIds.length);
-        for (uint i = 0; i < nonSignerOperatorIds.length; i++) {
-            uint192 nonSignerQuorumBitmap = 
-                registryCoordinator.getQuorumBitmapByOperatorIdAtBlockNumberByIndex(
-                    nonSignerOperatorIds[i], 
-                    referenceBlockNumber, 
-                    checkSignaturesIndices.nonSignerQuorumBitmapIndices[i]
-                );
-            // the number of quorums the operator was a part of that are also part of the provided quorumNumbers
-            checkSignaturesIndices.nonSignerStakeIndices[i] = new uint32[](BitmapUtils.countNumOnes(nonSignerQuorumBitmap & BitmapUtils.bytesArrayToBitmap(quorumNumbers)));
+        checkSignaturesIndices.nonSignerStakeIndices = new uint32[][](quorumNumbers.length);
+        for (uint8 quorumNumberIndex = 0; quorumNumberIndex < quorumNumbers.length; quorumNumberIndex++) {
+            uint256 numNonSignersForQuorum = 0;
+            // this array's length will be at most the number of nonSignerOperatorIds
+            checkSignaturesIndices.nonSignerStakeIndices[quorumNumberIndex] = new uint32[](nonSignerOperatorIds.length);
 
-            uint256 stakeIndexIndex = 0;
-            for (uint8 j = 0; j < 192; j++) {
+            for (uint i = 0; i < nonSignerOperatorIds.length; i++) {
+                uint192 nonSignerQuorumBitmap = 
+                    registryCoordinator.getQuorumBitmapByOperatorIdAtBlockNumberByIndex(
+                        nonSignerOperatorIds[i], 
+                        referenceBlockNumber, 
+                        checkSignaturesIndices.nonSignerQuorumBitmapIndices[i]
+                    );
+                
                 // if the operator was a part of the quorum and the quorum is a part of the provided quorumNumbers
-                if (nonSignerQuorumBitmap >> j & (BitmapUtils.bytesArrayToBitmap(quorumNumbers) >> j & 1) == 1) {
-                    checkSignaturesIndices.nonSignerStakeIndices[i][stakeIndexIndex] = stakeRegistry.getStakeUpdateIndexForOperatorIdForQuorumAtBlockNumber(
+                if (nonSignerQuorumBitmap >> uint8(quorumNumbers[quorumNumberIndex]) & 1 == 1) {
+                    checkSignaturesIndices.nonSignerStakeIndices[quorumNumberIndex][numNonSignersForQuorum] = stakeRegistry.getStakeUpdateIndexForOperatorIdForQuorumAtBlockNumber(
```

Comment:
are operator arrays trimmed to the right length using this, or do arrays end up all the same length with zeros filling them? I'm feeling like we end up with zeroes here which would be problematic in the event that we can't differentiate from when the correct stake index is zero?
maybe I'm just missing the high- to mid-level documentation on how all this signature checking logic is supposed to work / exactly how the inputs are supposed to be structured

Reply:
ahh I guess it all gets trimmed below. I think that works out then? again docs would help convince me that the design is right so I can then just check the design rather than trying to sort out the design from the code